packages/fuels-core/src/types/transaction_builders.rs

```diff
@@ -252,48 +245,139 @@ impl ScriptTransactionBuilder {
         ScriptTransactionBuilder {
             gas_price: None,
             gas_limit: None,
+            witness_limit: None,
+            max_fee: None,
             maturity: 0,
             script: vec![],
             script_data: vec![],
             inputs: vec![],
             outputs: vec![],
             witnesses: vec![],
             network_info,
+            gas_estimation_tolerance: 0.05,
             unresolved_signatures: Default::default(),
         }
     }
 
-    fn resolve_fuel_tx(self, base_offset: usize, num_witnesses: u8) -> Result<Script> {
-        let gas_price = self.gas_price.unwrap_or(self.network_info.min_gas_price);
-        let gas_limit = self.gas_limit.unwrap_or(self.network_info.max_gas_per_tx);
+    async fn build_with_provider(self, provider: &impl DryRunner) -> Result<ScriptTransaction> {
+        let is_using_predicates = self.is_using_predicates();
+        let base_offset = if is_using_predicates {
+            self.base_offset()
+        } else {
+            0
+        };
+
+        let num_witnesses = self.num_witnesses()?;
+        let tx = self
+            .resolve_fuel_tx_provider(base_offset, num_witnesses, provider)
+            .await?;
+
+        Ok(ScriptTransaction {
+            tx,
+            is_using_predicates,
+        })
+    }
+
+    fn create_dry_run_witnesses(&self, num_witnesses: u8) -> Vec<Witness> {
```

Comment:
A comment here explaining _why_ this is done the way it is would be nice for future-us.